### PR TITLE
feat: add support for custom error messages

### PR DESCRIPTION
**Pull request by tomer-yechiel**
implement custom error message
add tests
add documentation

### CODE REVIEW

**Comment by aiji42** [comment]
Hi @tomer-yechiel 
When setting up custom validation with `refine`, I think it would be simpler for users to translate messages directly using `t`.
For example,
```ts
 z.string().refine(() => false, i18next.t("custom_error_message") satisfies string)
```

I would like you to explain why you implemented `i18nKey` and more specific use cases.

**Comment by tomer-yechiel** [comment]
Hi @aiji42 
example of use case:
using i18next in server, and the local is derived from the request and the schema is define in other place, not inside the controller

**Comment by aiji42** [comment]
That use case example you presented could also be solved by using `t` directly as I wrote, right?

Since `{ params: { i18nKey: 'my_error_key' } }` does not work with type suggestion, I suspect that `i18nKey` will not be recalled in actual use cases, and most users will use `t` directly. (It seems simpler and makes more sense that way.)

Can you explain what your implemented feature does better than using `t` directly, or what we can't do without your feature?

**Comment by tomer-yechiel** [comment]
let say you have in the code a schema like the following 
` 
const schema = z.string().refine(() => false, { params: { i18nKey: "custom_error_message" } })
`

in another place in the code,  usually in a controller you could have something like this
```
controller(request){
  const t = createTfromRequest(request);
  
  await schema.safeParseAsync(formData, {errorMap: makeZodI18nMap({ t })});
}
```

i think its better then enforce the users to create the schema dynamically inside the controller

**Comment by aiji42** [comment]
Okay, it seems somewhat useful.
I just don't know what `createTfromRequest` does, so I doubt I understand what you mean.
As the owner of this library, I just want to understand when the functionality you have created works effectively.
Can you explain a little more what `createTfromRequest ` does? (not a detailed implementation, just an overview)

**Comment by tomer-yechiel** [comment]
createTfromRequest can create  t according to a local that is stored in a cookie

**Comment by aiji42** [comment]
I finally understand it. Thanks for the explanation.
I generally agree with the addition of this feature. Please wait for the code review.

File: packages/core/src/index.ts [hunk 1]
```diff
@@ -20,6 +20,7 @@ export type ZodI18nMapOption = {
   t?: i18n["t"];
   ns?: string | readonly string[];
   handlePath?: HandlePathOption;
+  customErrorNs?: string;
```
Comment (aiji42) [inline packages/core/src/index.ts:23]: There is no need to separate namespaces.
Multiple namespaces can be specified in `ns` as an array. `ns: ['zod', 'customErrorNS']`.

File: packages/core/src/index.ts [hunk 2]
```diff
@@ -213,11 +214,19 @@ export const makeZodI18nMap: MakeZodI18nMap = (option) => (issue, ctx) => {
...
+          defaultValue: message,
+          ...path,
+        });
+      } else {
+        message = t("errors.custom", {
+          ns,
+          defaultValue: message,
+          ...path,
+        });
+      }
```
Comment (aiji42) [inline packages/core/src/index.ts:229]:
```suggestion
      message = t(issue.params?.i18n || "errors.custom", {
        ns,
        defaultValue: message,
        ...path,
      });
```

File: packages/core/src/index.ts [hunk 3]
```diff
@@ -213,11 +214,19 @@ export const makeZodI18nMap: MakeZodI18nMap = (option) => (issue, ctx) => {
       );
       break;
     case ZodIssueCode.custom:
-      message = t("errors.custom", {
-        ns,
-        defaultValue: message,
-        ...path,
-      });
+      if (issue.params?.i18nKey !== undefined) {
```
Comment (aiji42) [inline packages/core/src/index.ts:217]: `i18nKey` is redundant, so use `i18n`.

**Review by aiji42 — CHANGES_REQUESTED** [review]

File: packages/core/src/index.ts (hunk 2 above)
Comment (tomer-yechiel) [reply packages/core/src/index.ts:229]: how this can work?
does i18next search in the all the namespaces until he find a namespace with the key?

Comment (aiji42) [reply packages/core/src/index.ts:229]: Yes, that's right.  
If `nsA` and `nsB` have the same key and the user wants to make the namespace explicit, the key can be `nsB:test_custom_key`.
This library has been created following the i18next specification as much as possible.

In fact, this test succeeds.
```ts
describe("custom error message", () => {
  test("custom error with refine", async () => {
    await i18next.init({
      lng: "en",
      resources: {
        en: {
          zod_custom: {
            test_custom_key: "custom error message",
          },
        },
      },
    });

    z.setErrorMap(makeZodI18nMap({ ns: ["zod", "zod_custom"] })); // <= multi ns

    expect(
      getErrorMessage(
        z
          .string()
          .refine(() => false, { params: { i18n: "test_custom_key" } })
          .safeParse("")
      )
    ).toEqual("custom error message");
  });
});
```

**Comment by tomer-yechiel** [comment]
@aiji42 done

**Review by aiji42 — APPROVED** [review]

**Comment by aiji42** [comment]
Thanks, I will merge and release this as soon as the test passes.

**Comment by aiji42** [comment]
@all-contributors please add @tomer-yechiel  for code and test

**Pull request merged**